File: ml-service/app/national_id.py
"""Egyptian national ID number: digit normalisation, structure validation and decoding.

Mirrors src/features/national-id/lib/nationalId.ts so the service and the web app agree.

    2  990101  21  0139  5
    |    |      |    |   +-- 14      check digit (algorithm not officially published)
    |    |      |    +------ 10-13   birth-registration sequence; the 13th digit is the gender
    |    |      |                    (odd = male, even = female)
    |    |      +----------- 8-9     governorate of birth (88 = born outside Egypt)
    |    +------------------ 2-7     birth date YYMMDD
    +----------------------- 1       century: 2 = 1900-1999, 3 = 2000-2099
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date

NATIONAL_ID_LENGTH = 14

GOVERNORATES: dict[str, tuple[str, str]] = {
    "01": ("Cairo", "القاهرة"),
    "02": ("Alexandria", "الإسكندرية"),
    "03": ("Port Said", "بورسعيد"),
    "04": ("Suez", "السويس"),
    "11": ("Damietta", "دمياط"),
    "12": ("Dakahlia", "الدقهلية"),
    "13": ("Sharqia", "الشرقية"),
    "14": ("Qalyubia", "القليوبية"),
    "15": ("Kafr El Sheikh", "كفر الشيخ"),
    "16": ("Gharbia", "الغربية"),
    "17": ("Monufia", "المنوفية"),
    "18": ("Beheira", "البحيرة"),
    "19": ("Ismailia", "الإسماعيلية"),
    "21": ("Giza", "الجيزة"),
    "22": ("Beni Suef", "بني سويف"),
    "23": ("Faiyum", "الفيوم"),
    "24": ("Minya", "المنيا"),
    "25": ("Asyut", "أسيوط"),
    "26": ("Sohag", "سوهاج"),
    "27": ("Qena", "قنا"),
    "28": ("Aswan", "أسوان"),
    "29": ("Luxor", "الأقصر"),
    "31": ("Red Sea", "البحر الأحمر"),
    "32": ("New Valley", "الوادي الجديد"),
    "33": ("Matrouh", "مطروح"),
    "34": ("North Sinai", "شمال سيناء"),
    "35": ("South Sinai", "جنوب سيناء"),
    "88": ("Born outside Egypt", "خارج الجمهورية"),
}

_CENTURY_START = {"2": 1900, "3": 2000}
_ARABIC_INDIC = str.maketrans("٠١٢٣٤٥٦٧٨٩۰۱۲۳۴۵۶۷۸۹", "01234567890123456789")


def normalize_digits(text: str) -> str:
    """Arabic-Indic (٠-٩) and Persian (۰-۹) digits -> 0-9."""
    return text.translate(_ARABIC_INDIC)


@dataclass(frozen=True)
class NationalIdInfo:
    id: str
    birth_date: date
    gender: str
    governorate_code: str
    governorate_en: str
    governorate_ar: str


def parse_national_id(raw: str, today: date | None = None) -> NationalIdInfo | str:
    """Returns decoded info, or an error message string explaining why the number is invalid."""
    today = today or date.today()
    digits = re.sub(r"[\s\-]", "", normalize_digits(raw))

    if not digits:
        return "National ID is required"
    if not digits.isdigit():
        return "National ID must contain digits only"
    if len(digits) != NATIONAL_ID_LENGTH:
        return f"National ID must be 14 digits (got {len(digits)})"

    century = _CENTURY_START.get(digits[0])
    if century is None:
        return f'Invalid century digit "{digits[0]}" (expected 2 or 3)'

    try:
        birth = date(century + int(digits[1:3]), int(digits[3:5]), int(digits[5:7]))
    except ValueError:
        return "Birth date encoded in the ID is not a valid date"
    if birth > today:
        return "Birth date encoded in the ID is in the future"

    code = digits[7:9]
    if code not in GOVERNORATES:
        return f'Unknown governorate code "{code}"'

    en, ar = GOVERNORATES[code]
    return NationalIdInfo(
        id=digits,
        birth_date=birth,
        gender="male" if int(digits[12]) % 2 == 1 else "female",
        governorate_code=code,
        governorate_en=en,
        governorate_ar=ar,
    )


def find_valid_id(digit_runs: list[str]) -> str | None:
    """First structurally valid 14-digit window across the given digit strings."""
    for run in digit_runs:
        digits = re.sub(r"\D", "", normalize_digits(run))
        for start in range(0, max(0, len(digits) - NATIONAL_ID_LENGTH) + 1):
            candidate = digits[start : start + NATIONAL_ID_LENGTH]
            if len(candidate) == NATIONAL_ID_LENGTH and isinstance(parse_national_id(candidate), NationalIdInfo):
                return candidate
    return None

File: src/features/national-id/lib/nationalId.ts
/**
 * Egyptian National ID number — structure
 * ---------------------------------------
 * Every Egyptian citizen has a 14-digit national number printed on the ID card.
 * The number is not random: most of it encodes facts about the holder.
 *
 *   Example:   2  900101  01  2345  6
 *              │    │      │    │   └─ [14]    Check digit (algorithm not officially published)
 *              │    │      │    └───── [10-13] Birth registration sequence.
 *              │    │      │                   The LAST of these 4 digits (13th overall) is the gender:
 *              │    │      │                   odd = male, even = female
 *              │    │      └────────── [8-9]   Governorate code of birth (88 = born abroad)
 *              │    └───────────────── [2-7]   Birth date as YYMMDD
 *              └────────────────────── [1]     Century of birth: 2 = 1900–1999, 3 = 2000–2099
 *
 * On the card the number is printed with Arabic-Indic digits (٠١٢٣٤٥٦٧٨٩),
 * so text coming from OCR must be normalised to 0-9 first.
 */

export const NATIONAL_ID_LENGTH = 14

/** Governorate codes (digits 8–9). */
export const GOVERNORATES: Record<string, { en: string; ar: string }> = {
  '01': { en: 'Cairo', ar: 'القاهرة' },
  '02': { en: 'Alexandria', ar: 'الإسكندرية' },
  '03': { en: 'Port Said', ar: 'بورسعيد' },
  '04': { en: 'Suez', ar: 'السويس' },
  '11': { en: 'Damietta', ar: 'دمياط' },
  '12': { en: 'Dakahlia', ar: 'الدقهلية' },
  '13': { en: 'Sharqia', ar: 'الشرقية' },
  '14': { en: 'Qalyubia', ar: 'القليوبية' },
  '15': { en: 'Kafr El Sheikh', ar: 'كفر الشيخ' },
  '16': { en: 'Gharbia', ar: 'الغربية' },
  '17': { en: 'Monufia', ar: 'المنوفية' },
  '18': { en: 'Beheira', ar: 'البحيرة' },
  '19': { en: 'Ismailia', ar: 'الإسماعيلية' },
  '21': { en: 'Giza', ar: 'الجيزة' },
  '22': { en: 'Beni Suef', ar: 'بني سويف' },
  '23': { en: 'Faiyum', ar: 'الفيوم' },
  '24': { en: 'Minya', ar: 'المنيا' },
  '25': { en: 'Asyut', ar: 'أسيوط' },
  '26': { en: 'Sohag', ar: 'سوهاج' },
  '27': { en: 'Qena', ar: 'قنا' },
  '28': { en: 'Aswan', ar: 'أسوان' },
  '29': { en: 'Luxor', ar: 'الأقصر' },
  '31': { en: 'Red Sea', ar: 'البحر الأحمر' },
  '32': { en: 'New Valley', ar: 'الوادي الجديد' },
  '33': { en: 'Matrouh', ar: 'مطروح' },
  '34': { en: 'North Sinai', ar: 'شمال سيناء' },
  '35': { en: 'South Sinai', ar: 'جنوب سيناء' },
  '88': { en: 'Born outside Egypt', ar: 'خارج الجمهورية' },
}

const CENTURY_START: Record<string, number> = { '2': 1900, '3': 2000 }

export interface NationalIdInfo {
  id: string
  birthDate: Date
  /** "YYYY-MM-DD", timezone independent. */
  birthDateIso: string
  age: number
  gender: 'male' | 'female'
  governorateCode: string
  governorate: { en: string; ar: string }
  sequence: string
  checkDigit: string
}

export type NationalIdResult = { ok: true; info: NationalIdInfo } | { ok: false; error: string }

/** Converts Arabic-Indic (٠-٩) and Eastern Arabic/Persian (۰-۹) digits to 0-9. */
export function normalizeDigits(text: string): string {
  return text
    .replace(/[٠-٩]/g, (d) => String(d.charCodeAt(0) - 0x0660))
    .replace(/[۰-۹]/g, (d) => String(d.charCodeAt(0) - 0x06f0))
}

/** Western digits → Arabic-Indic digits, the way they are printed on the card. */
export function toArabicDigits(text: string): string {
  return text.replace(/[0-9]/g, (d) => String.fromCharCode(0x0660 + Number(d)))
}

function calculateAge(birth: Date, today: Date): number {
  let age = today.getFullYear() - birth.getFullYear()
  const birthdayPassed =
    today.getMonth() > birth.getMonth() ||
    (today.getMonth() === birth.getMonth() && today.getDate() >= birth.getDate())
  if (!birthdayPassed) age -= 1
  return age
}

/**
 * Validates the structure of a national ID and decodes the information embedded in it.
 * Accepts Western or Arabic-Indic digits; spaces and dashes are ignored.
 */
export function parseNationalId(raw: string, today: Date = new Date()): NationalIdResult {
  const id = normalizeDigits(raw).replace(/[\s-]/g, '')

  if (!id) return { ok: false, error: 'National ID is required' }
  if (!/^\d+$/.test(id)) return { ok: false, error: 'National ID must contain digits only' }
  if (id.length !== NATIONAL_ID_LENGTH) {
    return { ok: false, error: `National ID must be 14 digits (got ${id.length})` }
  }

  const centuryDigit = id[0]
  const centuryStart = CENTURY_START[centuryDigit]
  if (centuryStart === undefined) {
    return { ok: false, error: `Invalid century digit "${centuryDigit}" (expected 2 or 3)` }
  }

  const year = centuryStart + Number(id.slice(1, 3))
  const month = Number(id.slice(3, 5))
  const day = Number(id.slice(5, 7))
  const birthDate = new Date(year, month - 1, day)
  // new Date() silently rolls over invalid dates (e.g. Feb 30 → Mar 2), so compare back.
  const isRealDate =
    birthDate.getFullYear() === year &&
    birthDate.getMonth() === month - 1 &&
    birthDate.getDate() === day
  if (!isRealDate) return { ok: false, error: 'Birth date encoded in the ID is not a valid date' }
  if (birthDate > today) return { ok: false, error: 'Birth date encoded in the ID is in the future' }

  const governorateCode = id.slice(7, 9)
  const governorate = GOVERNORATES[governorateCode]
  if (!governorate) return { ok: false, error: `Unknown governorate code "${governorateCode}"` }

  const sequence = id.slice(9, 13)
  const genderDigit = Number(id[12])

  return {
    ok: true,
    info: {
      id,
      birthDate,
      birthDateIso: `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`,
      age: calculateAge(birthDate, today),
      gender: genderDigit % 2 === 1 ? 'male' : 'female',
      governorateCode,
      governorate,
      sequence,
      checkDigit: id[13],
    },
  }
}
